test/javascripts/turbolinks_test.coffee

```diff
@@ -1,210 +1,946 @@
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadBCDA', ->
+            expectedLinkOrder = ['b.css', 'c.css', 'd.css', 'a.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 1) # Move A
+            done()
+
+      it 'maintains order when moving from a head with links ABCD to links CDAB.', (done) ->
+        visit url: 'fourLinksInHeadABCD', ->
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadCDAB', ->
+            expectedLinkOrder = ['c.css', 'd.css', 'a.css', 'b.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 2) # Move A,B or C,D
+            done()
+
+      it 'maintains order when moving from a head with links ABCD to links CDBA.', (done) ->
+        visit url: 'fourLinksInHeadABCD', ->
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadCDBA', ->
+            expectedLinkOrder = ['c.css', 'd.css', 'b.css', 'a.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 2) # Move A,B
+            done()
+
+      it 'maintains order when moving from a head with links ABCD to links CBDA.', (done) ->
+        visit url: 'fourLinksInHeadABCD', ->
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadCBDA', ->
+            expectedLinkOrder = ['c.css', 'b.css', 'd.css', 'a.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 2) # Move A,B or A,C
+            done()
+
+      it 'maintains order when moving from a head with links ABCD to links DCBA.', (done) ->
+        visit url: 'fourLinksInHeadABCD', ->
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadDCBA', ->
+            expectedLinkOrder = ['d.css', 'c.css', 'b.css', 'a.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 3) # Move any 3
+            done()
+
+  describe 'head script tag tracking', ->
```

Comment:
These should've been added after existing tests.  Comparison for `with partial page replacement` is pretty difficult now.